### PR TITLE
adi1090x-plymouth-themes: init at 2020-12-28

### PR DESCRIPTION
**Pull request by slwst**
###### Description of changes
* adi1090x-plymouth-themes: init at 2023-02-09

Provides Plymouth boot themes ported from android by adi1090x.  More info and theme previews available at [project upstream](https://github.com/adi1090x/plymouth-themes)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:13639]:
```suggestion
    keys = [{
      fingerprint = "6CEB 4A2F E6DC C345 1B2B  4733 AD52 C5FB 3EFE CC7A";
    }];
```

File: pkgs/data/themes/adi1090x-plymouth-themes/version.nix
Comment (SuperSandro2000) [inline pkgs/data/themes/adi1090x-plymouth-themes/version.nix:1]: Please inline this

File: pkgs/data/themes/adi1090x-plymouth-themes/update.sh
Comment (SuperSandro2000) [inline pkgs/data/themes/adi1090x-plymouth-themes/update.sh:4]: This needs to optionally take a GITHUB_TOKEN because we easily run into api rate limits

File: pkgs/data/themes/adi1090x-plymouth-themes/update.sh
Comment (SuperSandro2000) [inline pkgs/data/themes/adi1090x-plymouth-themes/update.sh:7]: This is wrong, the date should be the date of the commit, not when the script was rung

File: pkgs/data/themes/adi1090x-plymouth-themes/shas.nix
Comment (SuperSandro2000) [inline pkgs/data/themes/adi1090x-plymouth-themes/shas.nix:2]: If we are downloading so many files, why not download the entire repo and unpack the archives in it?

File: pkgs/data/themes/adi1090x-plymouth-themes/default.nix
Comment (SuperSandro2000) [inline pkgs/data/themes/adi1090x-plymouth-themes/default.nix:50]:
```suggestion
  stdenv.mkDerivation {
    pname = "adi1090x-plymouth-themes";
    inherit version srcs;

    nativeBuildInputs = [
      unzip
    ];

    sourceRoot = ".";
    unpackCmd = "tar xzf $curSrc";

    installPhase = ''
      mkdir -p $out/share/plymouth/themes
      for theme in ${toString selectedThemes}; do
        mv $theme $out/share/plymouth/themes/$theme
      done
      find $out/share/plymouth/themes/ -name \*.plymouth -exec sed -i "s@\/usr\/@$out\/@" {} \;
    '';

    meta = with lib; {
```
lets make that a bit nicer

File: pkgs/data/themes/adi1090x-plymouth-themes/shas.nix
Comment (slwst) [reply pkgs/data/themes/adi1090x-plymouth-themes/shas.nix:2]: Hi @SuperSandro2000!  Thanks for taking the time to review my PR 😃 This is my first contribution to NixOS and I'm really excited to be participating.

I would expect the typical user will only install one of these at a time as plymouth only supports a single boot animation to be configured.  If we download the entire repo the user will wind up with 500+mb of unused plymouth themes in their nix store.  This seems like unwanted behavior.

This approach was adapted from the nerdfonts pkg below as it has a similar problem (collection is huge, most users only want one or two).

https://github.com/NixOS/nixpkgs/tree/nixos-22.11/pkgs/data/fonts/nerdfonts

If this isn't an acceptable pattern I could shift to using a different output for each (similar to the catppuccin-cursors pkg), but the concern here is there are 80+ themes, that would be a lot of outputs...

File: pkgs/data/themes/adi1090x-plymouth-themes/update.sh
Comment (slwst) [reply pkgs/data/themes/adi1090x-plymouth-themes/update.sh]: Added this to update.sh and tested with a fine-grained access token with only the Public Repositories (read only) permission.

File: pkgs/data/themes/adi1090x-plymouth-themes/update.sh
Comment (slwst) [reply pkgs/data/themes/adi1090x-plymouth-themes/update.sh]: This has been set in `default.nix` to the upstream commit date.

File: pkgs/data/themes/adi1090x-plymouth-themes/shas.nix
Comment (SuperSandro2000) [reply pkgs/data/themes/adi1090x-plymouth-themes/shas.nix:2]: Oh, right. I was assuming we could just download the correct theme on demand but that wouldn't cache well. 

I just noticed another thing: all the archives are downloaded from master, so the hashes will be invalid if one gets updates.

**Comment by SuperSandro2000** [comment]
Please don't merge master into PRs. It makes rebasing and fixing merge conflicts unnecessary hard or sometimes near impossible.

File: pkgs/data/themes/adi1090x-plymouth-themes/shas.nix
Comment (slwst) [reply pkgs/data/themes/adi1090x-plymouth-themes/shas.nix:2]: Ahh great find.  Modified the script to retrieve the last master ref and make all calls using the ref-specific endpoints.  This should  guarantee that the shas.nix will always be accurate regardless if upstream has been updated and the package has not.

**Comment by SuperSandro2000** [comment]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits. I think after cleaning up the commit history we can merge this.

**Comment by slwst** [comment]
The commit history has been cleaned up in keeping with the contributing guide.  Thanks for your help through this process!

**Comment by slwst** [comment]
@SuperSandro2000 Is there something additional that needs to be done with this PR?

**Review by RaitoBezarius — APPROVED** [review]
Not a subject matter but LGTM.

**Comment by Quantumplation** [comment]
@slwst Sorry to hijack your PR, but do you have an example of how to use these themes? I'm a bit of a nix noob and can't quite figure it out; I saw you had a custom derivation for this in your own dotfiles, and it hasn't been updated to use this yet.

**Comment by slwst** [comment]
@Quantumplation Happy to help!  I just updated the `[slwst/dotfiles](https://github.com/slwst/dotfiles)` project to use the upstreamed `adi1090x-plymouth-themes` package.  Really all you need to do to use the themes is to configure plymouth as normal, then set `boot.plymouth.themePackages`  with the theme of your choosing.

Since the `adi1090x-plymouth-themes` package has quite a few themes in it, it defaults to a particular theme on install and allows this default to be overridden via selected_themes option.  Take a look at `hosts/shared/services/plymouth.nix` in the `[slwst/dotfiles](https://github.com/slwst/dotfiles)` repo for a usage example.  Take a look at `pkgs/data/themes/adi1090x-plymouth-themes/shas.nix` for a list of all possible theme names, and the original theme author's repo ( https://github.com/adi1090x/plymouth-themes ) for previews of them all.  

Good luck and feel free to ping me if you have any more questions!

**Pull request merged**
